File: frontend/src/api/subsidies.ts
/**
 * Subsidies registry API — реестр субсидий по компаниям портфеля.
 * Метрика «Субсидии» в финансах + модалка-реестр с фильтрами.
 */
import { api } from "./client";
import { fmtNumber } from "@/locale";
import { getCurrentLocale, t } from "@/locale/i18n";
import { i18nKey } from "@/locale/keys";

export interface SubsidyRow {
  id: string;
  company_id: string;
  company_name: string | null;
  company_code: string | null;
  sector_code: string | null;
  sector_name: string | null;
  sector_color: string | null;
  year: number | null;
  amount: number | null;
  program: string | null;
  source: string | null;
  kind: string | null;
  status: string | null;
  allocation_date: string | null;
  note: string | null;
  created_by_name: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface SubsidyUpsert {
  company_id: string;
  year?: number | null;
  amount?: number | null;
  program?: string | null;
  source?: string | null;
  kind?: string | null;
  status?: string | null;
  allocation_date?: string | null;
  note?: string | null;
}

export type SubsidyPatch = Omit<SubsidyUpsert, "company_id">;

export interface SubsidyCompanyAgg {
  company_id: string;
  company_name: string | null;
  company_code: string | null;
  sector_code: string | null;
  sector_name: string | null;
  sector_color: string | null;
  total: number;
  count: number;
}

export interface SubsidySectorAgg {
  sector_code: string | null;
  sector_name: string | null;
  sector_color: string | null;
  total: number;
  count: number;
}

export interface SubsidySummary {
  year: number | null;
  sector_code: string | null;
  total: number;
  count: number;
  by_company: SubsidyCompanyAgg[];
  by_sector: SubsidySectorAgg[];
}

export const subsidiesApi = {
  async list(params: { year?: number; sector_code?: string; company_id?: string } = {}): Promise<SubsidyRow[]> {
    const r = await api.get<SubsidyRow[]>("/subsidies", { params });
    return r.data;
  },

  async summary(params: { year?: number; sector_code?: string } = {}): Promise<SubsidySummary> {
    const r = await api.get<SubsidySummary>("/subsidies/summary", { params });
    return r.data;
  },

  async create(payload: SubsidyUpsert): Promise<SubsidyRow> {
    const r = await api.post<SubsidyRow>("/subsidies", payload);
    return r.data;
  },

  async update(id: string, patch: SubsidyPatch): Promise<SubsidyRow> {
    const r = await api.put<SubsidyRow>(`/subsidies/${id}`, patch);
    return r.data;
  },

  async remove(id: string): Promise<void> {
    await api.delete(`/subsidies/${id}`);
  },
};

// ─── Helpers ──────────────────────────────────────────────────────

/** Compact сум formatting: 1.5 трлн / 530 млрд / 12.0 млн сум. */
export function fmtSubsidySum(v: number | null | undefined): { value: string; unit: string } {
  if (v == null || isNaN(Number(v))) return { value: "—", unit: "" };
  const n = Number(v);
  const abs = Math.abs(n);
  if (abs >= 1e12) return { value: fmtNumber(n / 1e12, getCurrentLocale(), { decimals: 1 }), unit: t("трлн сум") };
  if (abs >= 1e9)  return { value: fmtNumber(n / 1e9, getCurrentLocale(), { decimals: 1 }),  unit: t("млрд сум") };
  if (abs >= 1e6)  return { value: fmtNumber(n / 1e6, getCurrentLocale(), { decimals: 1 }),  unit: t("млн сум") };
  if (abs >= 1e3)  return { value: fmtNumber(Math.round(n / 1e3), getCurrentLocale()), unit: t("тыс. сум") };
  return { value: fmtNumber(Math.round(n), getCurrentLocale()), unit: t("сум") };
}

export const SUBSIDY_STATUSES: { key: string; label: string }[] = [
  { key: "planned",   label: i18nKey("Запланирована") },
  { key: "allocated", label: i18nKey("Выделена") },
  { key: "received",  label: i18nKey("Получена") },
  { key: "used",      label: i18nKey("Освоена") },
  { key: "cancelled", label: i18nKey("Отменена") },
];

export function subsidyStatusLabel(key: string | null | undefined): string {
  if (!key) return "—";
  return SUBSIDY_STATUSES.find(s => s.key === key)?.label || key;
}

File: backend/app/schemas/subsidies.py
"""Subsidies registry schemas — реестр субсидий."""
from __future__ import annotations

from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas._types import MoneyDecimal


class SubsidyRow(BaseModel):
    """A single subsidy record (joined with company + sector for the registry)."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    company_name: Optional[str] = None
    company_code: Optional[str] = None
    sector_code: Optional[str] = None
    sector_name: Optional[str] = None
    sector_color: Optional[str] = None

    year: Optional[int] = None
    amount: Optional[MoneyDecimal] = None
    program: Optional[str] = None
    source: Optional[str] = None
    kind: Optional[str] = None
    status: Optional[str] = None
    allocation_date: Optional[date] = None
    note: Optional[str] = None

    created_by_name: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class SubsidyUpsert(BaseModel):
    """Create / update payload. company_id + year required on create."""
    company_id: UUID
    year: Optional[int] = None
    amount: Optional[float] = Field(None, ge=0)
    program: Optional[str] = Field(None, max_length=512)
    source: Optional[str] = Field(None, max_length=255)
    kind: Optional[str] = Field(None, max_length=128)
    status: Optional[str] = Field(None, max_length=32)
    allocation_date: Optional[date] = None
    note: Optional[str] = Field(None, max_length=8000)


class SubsidyPatch(BaseModel):
    """Partial update — все поля опциональны."""
    year: Optional[int] = None
    amount: Optional[float] = Field(None, ge=0)
    program: Optional[str] = Field(None, max_length=512)
    source: Optional[str] = Field(None, max_length=255)
    kind: Optional[str] = Field(None, max_length=128)
    status: Optional[str] = Field(None, max_length=32)
    allocation_date: Optional[date] = None
    note: Optional[str] = Field(None, max_length=8000)


class SubsidyCompanyAgg(BaseModel):
    company_id: UUID
    company_name: Optional[str] = None
    company_code: Optional[str] = None
    sector_code: Optional[str] = None
    sector_name: Optional[str] = None
    sector_color: Optional[str] = None
    total: MoneyDecimal
    count: int


class SubsidySectorAgg(BaseModel):
    sector_code: Optional[str] = None
    sector_name: Optional[str] = None
    sector_color: Optional[str] = None
    total: MoneyDecimal
    count: int


class SubsidySummary(BaseModel):
    """Aggregate for the «Субсидии» metric card (scope + year + sector filtered)."""
    year: Optional[int] = None
    sector_code: Optional[str] = None
    total: MoneyDecimal
    count: int
    by_company: list[SubsidyCompanyAgg] = Field(default_factory=list)
    by_sector: list[SubsidySectorAgg] = Field(default_factory=list)
